README.md
# easy-gd [![Build Status](https://travis-ci.org/furagu/easy-gd.png?branch=master)](https://travis-ci.org/furagu/easy-gd)

A simplified Node.js wrapper around [GD image manipulation library](http://libgd.bitbucket.org) with extra features:

* Image [format autodetection](#readingwriting-image-files): just `gd.open(file)` instead of choosing between gd.createFromJpeg() or gd.createFromPng() or whatever.
* Handy [resizing](#resizing-images) and [watermarking](#placing-a-watermark) shortcuts: `image.resize({width: 100, height:100})` and `image.watermark('logo.png')`.
* Reads/writes [files](#readingwriting-image-files), [buffers](#readingwriting-buffers) and [streams](#readingwriting-streams).
* Provides [synchronous](#synchronous-image-processing), [asynchronous](#asynchronous-image-processing) and [transform stream](#image-transform-streams) interfaces.
* Has built-in [Exif parsing](#reading-exif-data) and supports [automatic image orientation](#automatic-image-orientation).
* A drop-in replacement for [node-gd](https://www.npmjs.org/package/node-gd). You can just `require('easy-gd')` instead of `require('node-gd')` and everything will be working as before.


## Why should one use easy-gd?

Because it is fast. Easy-gd uses C++ buindings for libgd that add very little overhead. For example, it resizes images two times faster than [gm](https://www.npmjs.com/package/gm), which executes graphicsmagick commands in the background.


## Compatibility

Node.js 0.8-0.10.


## Installation

Ubuntu:

```bash
sudo apt-get install libgd2-xpm-dev
npm install easy-gd
```

Mac OS X:

```bash
brew install gd
npm install easy-gd
```


## Usage recipes

* [Resizing images](#resizing-images)
* [Placing a watermark](#placing-a-watermark)
* [Reading/writing image files](#readingwriting-image-files)
* [Reading/writing buffers](#readingwriting-buffers)
* [Reading/writing streams](#readingwriting-streams)
* [Image transform streams](#image-transform-streams)
* [Synchronous image processing](#synchronous-image-processing)
* [Asynchronous image processing](#asynchronous-image-processing)
* [Controlling the output format](#controlling-the-output-format)
* [Controlling image quality/compression](#controlling-image-qualitycompression)
* [Automatic filename extensions](#automatic-filename-extensions)
* [Reading Exif data](#reading-exif-data)
* [Automatic image orientation](#automatic-image-orientation)
* [Error handling](#error-handling)


### Resizing images

```js
var gd = require('easy-gd')

var image, resized

image = gd.open('image-800x600.jpg') // Source image size is 800×600

// Resize to feat into 100×100 box, yields 100×75 image
resized = image.resize({width: 100, height: 100})

// Resize by width, yields 200×150 image
resized = image.resize({width: 200})

// Resize by height, yields 267×200 image
resized = image.resize({height: 200})

// Resize and crop to 100×100
resized = image.crop({width: 100, height: 100})
resized = image.resize({width: 100, height: 100, method: 'crop'})

// Resize without resampling; faster but lowers the quality
resized = image.resize({width: 100, height: 100, resample: false})

// Save the resized image
resized.save('resized.jpg')
```

See also: [Reading/writing files](#readingwriting-image-files), [Asynchronous processing](#asynchronous-image-processing), [Image transform streams](#image-transform-streams).


### Placing a watermark

```js
var gd = require('easy-gd')

var image, watermarked

image = gd.open('source-image.jpg')

// Place a logo at the center of the image
watermarked = image.watermark('logo.png')
watermarked = image.watermark('logo.png', {x: 0.5, y: 0.5})

// At the left top corner
watermarked = image.watermark('logo.png', {x: 0, y: 0})

// At the right bottom corner
watermarked = image.watermark('logo.png', {x: 1, y: 1})

// Choose the most contrast position for a logo at the bottom
watermarked = image.watermark('logo.png', [{x: 0, y: 1}, {x: 0.5, y: 1}, {x: 1, y: 1}])

// Using gd.Image object as a watermark
var logo = gd.open('logo.png')
watermarked = image.watermark(logo)

// Save the watermarked image
watermarked.save('watermarked.jpg')
```

See also: [Reading/writing files](#readingwriting-image-files), [Reading/writing buffers](#readingwriting-buffers), [Reading/writing streams](#readingwriting-streams), [Asynchronous processing](#asynchronous-image-processing), [Error handling](#error-handling).


### Reading/writing image files

```js
var gd = require('easy-gd')

var image = gd.open('image.png')

// Do something to the image

image.save('processed.jpg', {quality: 90})
```

See also: [Reading/writing buffers](#readingwriting-buffers), [Reading/writing streams](#readingwriting-streams), [Asynchronous processing](#asynchronous-image-processing), [Controlling the output format](#controlling-the-output-format), [Error handling](#error-handling).


### Reading/writing buffers

```js
var gd = require('easy-gd')

// Reading image from buffer
var image = gd.open(imageData)

// Saving image to a buffer
var imageData = image.save()

// Using buffer as a watermark source
var watermarked = image.watermark(imageData)
```

See also: [Reading/writing files](#readingwriting-image-files), [Reading/writing streams](#readingwriting-streams), [Asynchronous processing](#asynchronous-image-processing), [Controlling the output format](#controlling-the-output-format), [Error handling](#error-handling).


### Reading/writing streams

```js
var gd = require('easy-gd')

// Reading image from stream
gd.open(stream, function (error, image) {
  if (!error) {
    // Process the image
  }
})

// Saving image to a stream
image.save(stream, function (error) {
  if (!error) {
    // Some action
  }
})

// Using stream as a watermark source
image.watermark(stream, function (error, watermarked) {
  if (!error) {
    // Process the watermarked image
  }
})
```

See also: [Image transform streams](#image-transform-streams), [Reading/writing files](#readingwriting-image-files), [Reading/writing buffers](#readingwriting-buffers), [Controlling the output format](#controlling-the-output-format), [Error handling](#error-handling).


### Image transform streams

All the image manipulation methods called directly on the module object produce chainable [transform streams](http://nodejs.org/api/stream.html#stream_class_stream_transform_1):

```js
var gd = require('easy-gd')

// Making thumbnails
process.stdin
  .pipe(gd.crop({width: 100, height: 100}))
  .pipe(process.stdout)

// Watermarking
process.stdin
  .pipe(gd.watermark('logo.png'))
  .pipe(process.stdout)

// Changing image format
process.stdin
  .pipe(gd.format('jpeg').quality(90))
  .pipe(process.stdout)

// Combine everything
process.stdin
  .pipe(
    gd.resize({width: 800, height: 600})
      .watermark('logo.png', {x:1, y:1})
      .options({format: 'jpeg', quality: '90'})
  )
  .pipe(process.stdout)
```

See also: [Reading/writing files](#readingwriting-image-files), [Reading/writing buffers](#readingwriting-buffers), [Reading/writing streams](#readingwriting-streams), [Controlling the output format](#controlling-the-output-format), [Error handling](#error-handling).


### Synchronous image processing

With easy-gd you can synchronously process files and buffers:

```js
var gd = require('easy-gd')

// Processing files
gd.open('input.png')
  .resize({width: 800, height: 600})
  .save('output.jpg', {quality: 90})

// Processing buffers
var outputData = gd.open(inputData)
  .resize({width: 800, height: 600})
  .save({format: 'jpeg', quality: 90})
```

See also: [Asynchronous processing](#asynchronous-image-processing), [Error handling](#error-handling).


### Asynchronous image processing

You can asynchronously process files, buffers and streams by passing additional `callback` argument:

```js
var gd = require('easy-gd')

// Processing files
gd.open('input.png', function (error, image) {
  if (error) throw error
  image.resize({width: 800, height: 600}, function (error, resized) {
    if (error) throw error
    resized.save('output.jpg', {quality: 90}, function (error) {
      if (error) throw error
    })
  })
})

// Processing buffers
gd.open(inputData, function (error, image) {
  if (error) throw error
  image.resize({width: 800, height: 600}, function (error, resized) {
    if (error) throw error
    resized.save({format: 'jpeg', quality: 90}, function (error, outputData) {
      if (error) throw error
      // Process outputData buffer
    })
  })
})

// Processing streams
gd.open(inputStream, function (error, image) {
  if (error) throw error
  image.resize({width: 800, height: 600}, function (error, resized) {
    if (error) throw error
    resized.save(outputStream, {format: 'jpeg', quality: 90}, function (error) {
      if (error) throw error
    })
  })
})
```

See also: [Image transform streams](#image-transform-streams), [Synchronous processing](#synchronous-image-processing), [Error handling](#error-handling).


### Controlling the output format

```js
var gd = require('easy-gd')

// Format is inherited from the source image
var image = gd.open('input.jpg')
var resizedBuffer = image.resize({width: 100}).save() // Saved in JPEG

// Format can be specified explicitly with target filename extension
var image = gd.open('input.jpg')
image.save('output.png') // Saved in PNG

// Format can be specified explicitly with save({format: 'format_name'})
var image = gd.open('input.jpg')
var pngBuffer = image.save({format: 'png'}) // Saved in PNG

// Target file extension has higher priority
var image = gd.open('input.png')
image.save('output.jpg') // Saved in JPEG
```

Format specification priority: filename extension > save 'format' option > inherited format.

See also: [Controlling image quality/compression](#controlling-image-qualitycompression), [Automatic filename extensions](#automatic-filename-extensions), [Error handling](#error-handling).


### Controlling image quality/compression

```js
var gd = require('easy-gd')
var image = gd.open('input.jpg')

// Setting JPEG file quality, 0-100
image.save('output.jpg', {quality: 90})

// Setting PNG file compression level, 0-9
image.save('ouput.png', {compression: 6})

// Transform stream options
inputStream.pipe(gd.format('png').compression(6)).pipe(outputStream)
inputStream.pipe(gd.format('jpeg').quality(90)).pipe(outputStream)
inputStream.pipe(gd.options({format: 'jpeg', quality: 90}).pipe(outputStream)

// Buffer saving options
var outputBuffer = image.save({format: 'jpeg', quality: 90})
var outputBuffer = image.save({format: 'png', compression: 6})
```

See also: [Controlling the output format](#controlling-the-output-format), [Automatic filename extensions](#automatic-filename-extensions).


### Automatic filename extensions

```js
var gd = require('easy-gd')

var image = gd.open('input.jpg')
image.save('output.{ext}', {format: 'png'}) // Writes ouput.png


var image = gd.open('input.jpg')
image
  .resize({width: 100})
  .save('output.{ext}') // Writes output.jpg since format was inherited
```

See also: [Controlling the output format](#controlling-the-output-format), [Controlling image quality/compression](#controlling-image-qualitycompression).


### Reading Exif data

Exif data are being parsed automatically for JPEG images.

```js
var gd = require('easy-gd')

var image = gd.open('input.jpg')

// Accessing Exif tags
if (image.exif) {
  console.log('%s %s', image.exif.GPSLongitude, image.exif.GPSLatitude)
} else {
  console.log('No Exif data')
}
```

Note: image.exif property will be copied by resize() and other methods, but will not be written to the destination image.

See also: [Automatic image orientation](#automatic-image-orientation).


### Automatic image orientation

GD does not process Exif data, resulting rotated images in the output. Easy-gd fixes this by automatically orienting the image.

```js
var gd = require('easy-gd')

// The image gets automatically oriented by Exif data
var image = gd.open('photo.jpg')

// Turn automatic orientation off
var original = gd.open('photo.jpg', {autoOrient: false})

// Automatically orient existing image
var rotated = original.autoOrient()
```

See also: [Reading Exif data](#reading-exif-data).


### Error handling

```js
var gd = require('easy-gd')

// Synchronous methods throw exceptions
try {
  var image = gd.open('non-existent.png')
} catch (error) {
  console.log('Failed to open the image: %s', error)
}

// Asynchronous methods return errors as the first callback argument;
// null means there was no error
gd.open('non-existent.png', function (error, image) {
  if (error) {
    console.log('Failed to open the image: %s', error)
    return
  }
  // ...
})

// Every error raised or returned by easy-gd is a descendant or gd.Error
try {
  doImageRelatedStuff()
} catch (error) {
  if (error instanceof gd.Error) {
    console.log('Image processing error: %s', error)
  } else {
    // Some other error happened
  }
}
```

There are some subclasses you can use to catch specific errors:

* __gd.UnknownSourceType__ - unknown source type.
* __gd.EmptySource__ - empty source file or buffer.
* __gd.UnknownImageFormat__ - unknown image format (or not an image at all).
* __gd.IncompleteImage__ - corrupted or incomplete image.
* __gd.UnsupportedOrientation__ - unsupported image Exif orientation tag value.
* __gd.DestinationFormatRequired__ - destination image format required.
* __gd.UnknownDestinationType__ - unknown destination type.
* __gd.FileOpen__ - file opening error.
* __gd.FileDoesNotExist__ - file does not exist.
* __gd.FileWrite__ - file writing error.
* __gd.SynchronousStreamAccess__ - a stream cannot be read or written synchronously.
* __gd.OptionsRequired__ - options argument should be passed in.


CHANGELOG
1.0.0
- Improved the interface consistency
- Added transform streams support

0.0.7
- fixed multiple callback issue https://github.com/furagu/easy-gd/issues/1
- added `options` argument to gd.createFrom()
- switched from exif to exif-parser lib
- added tests for autorotate feature


test.js
var gd = require('./index.js')

gd.createTrueColor(100, 100).saveJpeg('test.jpg')

gd.createFrom('test.jpg', function (err, image) {
    console.log(err, image.width, image.height)
})


test/ptr.js
var should = require('should'),
    gd = require('../index.js'),
    _ = require('underscore'),
    testData = require('./data.js')

describe('gd', function () {
    describe('getFormatPtr()', function () {
        testData.forBufferType(function (buffer, type) {
            it('should detect ' + type, function () {
                gd.getFormatPtr(buffer).should.equal(type)
            })
        })
        it('should throw unknown_format error on bad data', function () {
            _.partial(gd.getFormatPtr, new Buffer('BADDATA')).should.throw('Unknown image format')
        })
    })

    describe('createFromPtr()', function () {
        testData.forBufferType(function (buffer, type) {
            it('should open ' + type + ' buffer', function () {
                var image = gd.createFromPtr(buffer)
                image.should.be.an.instanceof(gd.Image)
                image.width.should.equal(1)
                image.height.should.equal(1)
                image.format.should.equal(type)
            })
        })
        it('should throw unknown_format error on bad data', function () {
            _.partial(gd.getFormatPtr, new Buffer('BADDATA')).should.throw('Unknown image format')
        })
    })

    describe('Image.prototype', function () {
        var testImage = gd.createTrueColor(100, 100)
        testImage.filledEllipse(50, 50, 25, 25, testImage.colorAllocate(255, 0, 0))

        describe('targetFormat()', function () {
            it('should take format from options.format', function () {
                testImage.targetFormat({format: 'FORMAT'}).should.equal('FORMAT')
            })
            it('should take format from options.defaultFormat', function () {
                testImage.targetFormat({defaultFormat: 'FORMAT'}).should.equal('FORMAT')
            })
            it('should take format from this.format', function () {
                _.extend(testImage, {format: 'FORMAT'}).targetFormat().should.equal('FORMAT')
            })
            it('should prefer options.format over this.format and options.defaultFormat', function () {
                _.extend(testImage, {format: 'THISFORMAT'})
                    .targetFormat({format: 'FORMAT', defaultFormat: 'DEFAULTFORMAT'}).should.equal('FORMAT')
            })
            it('should prefer this.format over options.defaultFormat', function () {
                _.extend(testImage, {format: 'THISFORMAT'})
                    .targetFormat({defaultFormat: 'DEFAULTFORMAT'}).should.equal('THISFORMAT')
            })
            delete testImage.format
        })

        describe('ptr()', function () {
            var gdOpeners = {
                    'jpeg': gd.createFromJpegPtr,
                    'png':  gd.createFromPngPtr,
                    'gif':  gd.createFromGifPtr,
                }

            testData.forType(function (type) {
                it('should create a ' + type + ' image', function () {
                    var image = gdOpeners[type](testImage.ptr({format: type}))
                    image.should.be.an.instanceof(gd.Image)
                    image.width.should.equal(testImage.width)
                    image.height.should.equal(testImage.height)
                })
            })
            it('should set jpeg images quality with jpegquality option', function () {
                var highQualityImage = testImage.ptr({format: 'jpeg', jpegquality: 99}),
                    lowQualityImage  = testImage.ptr({format: 'jpeg', jpegquality: 1})
                lowQualityImage.length.should.be.below(highQualityImage.length)
            })
            it('should compress png images with pnglevel option', function () {
                var highlyCompressedImage = testImage.ptr({format: 'png', pnglevel: 9}),
                    littleCompressedImage = testImage.ptr({format: 'png', pnglevel: 1})
                highlyCompressedImage.length.should.be.below(littleCompressedImage.length)
            })
        })
    })
})


test/data.js
var _ = require('underscore'),
    testdata = {
        'gif':  new Buffer([0x47, 0x49, 0x46, 0x38, 0x37, 0x61, 0x1, 0x0, 0x1, 0x0, 0x80, 0x0, 0x0, 0xfc, 0x2, 0x4, 0x0, 0x0, 0x0, 0x2c, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1, 0x0, 0x0, 0x2, 0x2, 0x44, 0x1, 0x0, 0x3b]),
        'jpeg': new Buffer([0xff, 0xd8, 0xff, 0xe0, 0x0, 0x10, 0x4a, 0x46, 0x49, 0x46, 0x0, 0x1, 0x1, 0x0, 0x0, 0x1, 0x0, 0x1, 0x0, 0x0, 0xff, 0xfe, 0x0, 0x3c, 0x43, 0x52, 0x45, 0x41, 0x54, 0x4f, 0x52, 0x3a, 0x20, 0x67, 0x64, 0x2d, 0x6a, 0x70, 0x65, 0x67, 0x20, 0x76, 0x31, 0x2e, 0x30, 0x20, 0x28, 0x75, 0x73, 0x69, 0x6e, 0x67, 0x20, 0x49, 0x4a, 0x47, 0x20, 0x4a, 0x50, 0x45, 0x47, 0x20, 0x76, 0x39, 0x30, 0x29, 0x2c, 0x20, 0x71, 0x75, 0x61, 0x6c, 0x69, 0x74, 0x79, 0x20, 0x3d, 0x20, 0x31, 0x30, 0x30, 0xa, 0xff, 0xdb, 0x0, 0x43, 0x0, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0xff, 0xdb, 0x0, 0x43, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0xff, 0xc0, 0x0, 0x11, 0x8, 0x0, 0x1, 0x0, 0x1, 0x3, 0x1, 0x22, 0x0, 0x2, 0x11, 0x1, 0x3, 0x11, 0x1, 0xff, 0xc4, 0x0, 0x1f, 0x0, 0x0, 0x1, 0x5, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0xa, 0xb, 0xff, 0xc4, 0x0, 0xb5, 0x10, 0x0, 0x2, 0x1, 0x3, 0x3, 0x2, 0x4, 0x3, 0x5, 0x5, 0x4, 0x4, 0x0, 0x0, 0x1, 0x7d, 0x1, 0x2, 0x3, 0x0, 0x4, 0x11, 0x5, 0x12, 0x21, 0x31, 0x41, 0x6, 0x13, 0x51, 0x61, 0x7, 0x22, 0x71, 0x14, 0x32, 0x81, 0x91, 0xa1, 0x8, 0x23, 0x42, 0xb1, 0xc1, 0x15, 0x52, 0xd1, 0xf0, 0x24, 0x33, 0x62, 0x72, 0x82, 0x9, 0xa, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2a, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3a, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48, 0x49, 0x4a, 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7a, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89, 0x8a, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97, 0x98, 0x99, 0x9a, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9, 0xaa, 0xb2, 0xb3, 0xb4, 0xb5, 0xb6, 0xb7, 0xb8, 0xb9, 0xba, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7, 0xc8, 0xc9, 0xca, 0xd2, 0xd3, 0xd4, 0xd5, 0xd6, 0xd7, 0xd8, 0xd9, 0xda, 0xe1, 0xe2, 0xe3, 0xe4, 0xe5, 0xe6, 0xe7, 0xe8, 0xe9, 0xea, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9, 0xfa, 0xff, 0xc4, 0x0, 0x1f, 0x1, 0x0, 0x3, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0xa, 0xb, 0xff, 0xc4, 0x0, 0xb5, 0x11, 0x0, 0x2, 0x1, 0x2, 0x4, 0x4, 0x3, 0x4, 0x7, 0x5, 0x4, 0x4, 0x0, 0x1, 0x2, 0x77, 0x0, 0x1, 0x2, 0x3, 0x11, 0x4, 0x5, 0x21, 0x31, 0x6, 0x12, 0x41, 0x51, 0x7, 0x61, 0x71, 0x13, 0x22, 0x32, 0x81, 0x8, 0x14, 0x42, 0x91, 0xa1, 0xb1, 0xc1, 0x9, 0x23, 0x33, 0x52, 0xf0, 0x15, 0x62, 0x72, 0xd1, 0xa, 0x16, 0x24, 0x34, 0xe1, 0x25, 0xf1, 0x17, 0x18, 0x19, 0x1a, 0x26, 0x27, 0x28, 0x29, 0x2a, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3a, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48, 0x49, 0x4a, 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7a, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89, 0x8a, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97, 0x98, 0x99, 0x9a, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9, 0xaa, 0xb2, 0xb3, 0xb4, 0xb5, 0xb6, 0xb7, 0xb8, 0xb9, 0xba, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7, 0xc8, 0xc9, 0xca, 0xd2, 0xd3, 0xd4, 0xd5, 0xd6, 0xd7, 0xd8, 0xd9, 0xda, 0xe2, 0xe3, 0xe4, 0xe5, 0xe6, 0xe7, 0xe8, 0xe9, 0xea, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9, 0xfa, 0xff, 0xda, 0x0, 0xc, 0x3, 0x1, 0x0, 0x2, 0x11, 0x3, 0x11, 0x0, 0x3f, 0x0, 0xfc, 0x5f, 0xa2, 0x8a, 0x2b, 0xfc, 0xa7, 0x3f, 0xef, 0xe0, 0xff, 0xd9]),
        'png':  new Buffer([0x89, 0x50, 0x4e, 0x47, 0xd, 0xa, 0x1a, 0xa, 0x0, 0x0, 0x0, 0xd, 0x49, 0x48, 0x44, 0x52, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x1, 0x8, 0x2, 0x0, 0x0, 0x0, 0x90, 0x77, 0x53, 0xde, 0x0, 0x0, 0x0, 0xc, 0x49, 0x44, 0x41, 0x54, 0x8, 0x99, 0x63, 0xf8, 0xcf, 0xc0, 0x0, 0x0, 0x3, 0x1, 0x1, 0x0, 0x9c, 0xe3, 0xbf, 0x59, 0x0, 0x0, 0x0, 0x0, 0x49, 0x45, 0x4e, 0x44, 0xae, 0x42, 0x60, 0x82]),
    },
    filename = __dirname + 'testdata'

exports.forFileData = function (callback, done) {
    _.each(testdata, function (data, type) {
        callback(filename + '.' + type, data)
    })
    if (done) done()
}

exports.forFileType = function (callback, done) {
    _.each(testdata, function (data, type) {
        callback(filename + '.' + type, type)
    })
    if (done) done()
}

exports.forBufferType = function (callback, done) {
    _.each(testdata, callback)
    if (done) done()
}

exports.forType = function (callback, done) {
    _.each(testdata, function (data, type) {
        callback(type)
    })
    if (done) done()
}


lib/errors.js
var util = require('util')

var errorsDefinition = [
    ['UnknownSourceType',           'Unknown source type.'],
    ['EmptySource',                 'Empty source file or buffer.'],
    ['UnknownImageFormat',          'Unknown image format (or not an image at all).'],
    ['IncompleteImage',             'Corrupted or incomplete image.'],
    ['UnsupportedOrientation',      'Unsupported image Exif orientation tag value.'],
    ['DestinationFormatRequired',   'Destination image format required.'],
    ['UnknownDestinationType',      'Unknown destination type.'],
    ['FileOpen',                    'File open error.'],
    ['FileDoesNotExist',            'File does not exist.'],
    ['FileWrite',                   'File writing error.'],
    ['SynchronousStreamAccess',     'A stream cannot be read or written synchronously.'],
    ['OptionsRequired',             'An options argument should be passed in.'],
]

exports.Error = errorClass('Error', Error)
exports.classes = {'Error': Error}

errorsDefinition.forEach(function (definition) {
    var className = definition[0] + 'Error',
        defaultMessage = definition[1]
    exports[className] = errorClass(className, exports.Error, defaultMessage)
    exports.classes[className] = exports[className]
})

exports.routeError = function routeError(error, callback) {
    if (callback) return callback(error)
    throw error
}

function errorClass(name, parentClass, defaultMessage) {
    function CustomError(message) {
        if (!(this instanceof CustomError)) return new CustomError(message)
        this.name = name
        this.message = message || defaultMessage
        Error.captureStackTrace(this, CustomError)
    }
    util.inherits(CustomError, parentClass)
    return CustomError
}


lib/opener.js

var fs = require('fs'),
    buffertools = require('buffertools'),
    stream = require('readable-stream'),
    vargs = require('vargs-callback'),
    _ = require('underscore'),
    formats = require('./formats.js'),
    errors = require('./errors.js')

var openDefaults = {
    autoOrient: true,
}

exports.open = vargs(function open(source, options, callback) {
    var async = !!callback
    return readSource(source, async, function (err, imageData) {
        if (err) return errors.routeError(err, callback)
        try {
            var image = openImage(imageData, options)
            if (async) return callback(null, image)
            return image
        } catch (e) {
            return errors.routeError(e, callback)
        }
    })
})

function readSource(source, async, callback) {
    if (source instanceof Buffer) return callback(null, source)

    if (typeof source === 'string') {
        if (async) return readStream(fs.createReadStream(source), callback)
        try {
            var data = fs.readFileSync(source)
        } catch (e) {
            return wrapFileReadingError(e, callback)
        }
        return callback(null, data)
    }

    if (source.on) {
        if (!async) throw errors.SynchronousStreamAccessError()
        return readStream(source, callback)
    }

    return callback(errors.UnknownSourceTypeError())
}

function openImage(imageData, options) {
    options = _(options || {}).defaults(openDefaults)

    if (!imageData.length) throw errors.EmptySourceError()

    var format = formats.getImageFormat(imageData)
    var image = format.open(imageData)
    if (!image) throw errors.IncompleteImageError()
    image.format = format.label

    if (options.autoOrient && image.exif) {
        try {
            return image.autoOrient()
        } catch (e) {
            // ignore auto orientation errors
        }
    }

    return image
}

function readStream(stream, callback) {
    var dataRead = Buffer(0)

    stream.on('data', function (chunk) {
        dataRead = buffertools.concat(dataRead, chunk)
    })

    stream.on('end', function () {
        return callback(null, dataRead)
    })

    stream.on('error', function (error) {
        return wrapFileReadingError(error, callback)
    })
}

function wrapFileReadingError(error, callback) {
    if (error.code === 'ENOENT') return callback(errors.FileDoesNotExistError())
    return callback(errors.FileOpenError(error.message))
}


lib/transform.js
var gd = require('../index.js'),
    buffertools = require('buffertools'),
    stream = require('readable-stream'),
    util = require('util'),
    async = require('async'),
    _ = require('underscore')


function ImageTransformStream(options) {
    if (!(this instanceof ImageTransformStream)) return new ImageTransformStream(options)
    stream.Transform.call(this, options)
    this.imageData = Buffer(0)
    this.imageOptions = {}
    this.actions = []
}
util.inherits(ImageTransformStream, stream.Transform)

ImageTransformStream.prototype._transform = function collectImageData(chunk, encoding, done) {
    this.imageData = buffertools.concat(this.imageData, chunk)
    done()
}

ImageTransformStream.prototype._flush = function processImage(done) {
    this.actions.unshift(_.partial(gd.open, this.imageData))
    this.actions.push(function saveImage(image, callback) {
        var processedImageData = image.save(this.imageOptions)
        this.push(processedImageData)
        callback()
    }.bind(this))
    async.waterfall(this.actions, done)
}

var methods = {}

methods.options = function options(options) {
    _.extend(this.imageOptions, options)
    return this
}

_.each(['format', 'quality', 'compression'], function (option) {
    methods[option] = function (optionValue) {
        this.imageOptions[option] = optionValue
        return this
    }
})

_.each(['resize', 'watermark', 'crop'], function (action) {
    methods[action] = function () {
        var args = Array.prototype.slice.apply(arguments)
        this.actions.push(function (image, callback) {
            image[action].apply(image, args.concat([callback]))
        })
        return this
    }
})

_.extend(ImageTransformStream.prototype, methods)

_.each(methods, function (method, name) {
    exports[name] = function () {
        return method.apply(ImageTransformStream(), arguments)
    }
})


lib/formats.js
var path = require('path'),
    buffertools = require('buffertools'),
    exifParser = require('exif-parser'),
    _ = require('underscore'),
    gd = require('node-gd'),
    errors = require('./errors.js')

var formats = [{
        label: 'jpeg',
        ext: 'jpg',
        signature: Buffer([0xff, 0xd8, 0xff]),
        open: function openJpeg(imageData) {
            var image = gd.createFromJpegPtr(imageData)
            try {
                var exif = exifParser.create(imageData).parse()
            } catch (e) {
                // ignore exif parsing errors
            }
            if (!_.isEmpty(exif.tags)) image.exif = exif.tags
            return image
        },
        save: function saveJpeg(image, options) {
            return image.jpegPtr(options.quality || -1)
        },
    }, {
        label: 'png',
        ext: 'png',
        signature: Buffer([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
        open: gd.createFromPngPtr,
        save: function savePng(image, options) {
            return image.pngPtr(options.compression || -1)
        },
    }, {
        label: 'gif',
        ext: 'gif',
        signature: Buffer('GIF'),
        open: gd.createFromGifPtr,
        save: function saveGif(image) {
            return image.gifPtr()
        },
    }
]

exports.getImageFormat = function getImageFormat(imageData) {
    var format = _.find(formats, function (f) {
        return buffertools.equals(imageData.slice(0, f.signature.length), f.signature)
    })
    if (format) return format
    throw errors.UnknownImageFormatError()
}

var formatsByLabel = _.object(_.map(formats, function (f) {return [f.label, f]}))
var formatsByExtname = _.object(_.map(formats, function (f) {return ['.' + f.ext, f]}))
formatsByExtname['.jpeg'] = formatsByExtname['.jpg']

exports.getDestinationFormat = function getDestinationFormat(image, options, filename) {
    if (filename) {
        var extname = path.extname(filename).toLowerCase()
        if (extname in formatsByExtname) return formatsByExtname[extname]
    }
    var label = options.format || image.format
    if (label) {
        if (label in formatsByLabel) return formatsByLabel[label]
        throw errors.UnknownImageFormatError('Unknown format ' + label)
    }
    throw errors.DestinationFormatRequiredError()
}
